File: 76.minimum-window-substring.java
/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */
class Solution {
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.length() < t.length()) return "";
        int[] freq = new int[256];
        int[] save = new int[256];
        for (int i = 0; i < t.length(); ++i) freq[t.charAt(i)]++;
        for (int i = 0; i < freq.length; ++i) if (freq[i] == 0) freq[i] = -1;
        int i = 0, j = -1, ai = 0, aj = s.length();
        int charLeft = t.length();
        
        while (i < s.length()) {
            if (charLeft > 0 && j < s.length() - 1) {

                if (freq[s.charAt(++j)] > 0) {
                    freq[s.charAt(j)]--;
                    charLeft--;
                }
                else if (freq[s.charAt(j)] == 0)
                    save[s.charAt(j)]++;

            }
            else {

                if (charLeft == 0 && j-i+1 < aj-ai+1) {
                    ai = i;
                    aj = j;
                }

                if (save[s.charAt(i)] > 0) {
                    save[s.charAt(i)]--;
                }
                else if (freq[s.charAt(i)] == 0) {
                    if (j == s.length() - 1) break;
                    freq[s.charAt(i)]++;
                    charLeft++;
                }

                i++;
            }
        }
        return aj == s.length() ? "" : s.substring(ai, aj+1);
    }
}

File: go/76.minimum-window-substring.go
package main

/*
 * @lc app=leetcode id=76 lang=golang
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
func minWindow(s string, t string) string {
	if len(t) > len(s) {
		return ""
	}
	var cp, mem [256]int
	for i := 0; i < len(t); i++ {
		cp[int(t[i])]++
		mem[int(s[i])]++
	}
	var res string
	for j, i := 0, len(t)-1; ; {
		if isContaining(&mem, &cp) {
			if len(res) == 0 || len(res) > i+1-j {
				res = s[j : i+1]
			}
			mem[int(s[j])]--
			j++
		} else {
			i++
			if i >= len(s) {
				break
			}
			mem[int(s[i])]++
		}
	}
	return res
}

func isContaining(mem *[256]int, cp *[256]int) bool {
	for i := 0; i < 256; i++ {
		if (*mem)[i] < (*cp)[i] {
			return false
		}
	}
	return true
}

// @lc code=end
